### PR TITLE
[build] Remove coqmktop in favor of ocamlfind.

### PR DESCRIPTION
**Pull request by ejgallego**
We remove coqmktop in favor of a couple of simple makefile rules using
ocamlfind. In order to do that, we introduce a new top-level file that
calls the coqtop main entry.

This is very convenient in order to use other builds systems such as
`ocamlbuild` or `jbuilder`.

An additional consideration is that we must perform a side-effect on
init depending on whether we have an OCaml toplevel available [byte]
or not. We do that by using two different object files, one for the
bytecode version other for the native one, but we may want to review
our choice.

We also perform some smaller cleanups taking profit from ocamlfind.

### CODE REVIEW

**Comment by ejgallego** [comment]
Failure here is interesting:
```
File "lib/lib.cma(Profile)", line 1:
Error (warning 31): files lib/lib.cma(Profile) and /home/travis/.opam/4.06.0+trunk/lib/ocaml/compiler-libs/ocamlcommon.cma(Profile) both define a module named Profile
```
due to https://github.com/ocaml/ocaml/pull/1152 , we don't see that in the old build mode due to less toplevel modules being linked in. I wonder what's the deal here.

File: Makefile.build [hunk 1]
```diff
@@ -379,32 +377,40 @@ coqbinaries: $(COQMKTOP) $(COQTOPEXE) $(CHICKEN) $(CSDPCERT) $(FAKEIDE)
...
+	$(HIDE)$(OCAMLC) $(MLINCLUDES) $(BYTEFLAGS) -package compiler-libs.toplevel -c $<
+
 ifeq ($(BEST),opt)
-$(COQTOPEXE): $(COQMKTOP) $(LINKCMX) $(LIBCOQRUN) $(TOPLOOPCMA:.cma=.cmxs)
+$(COQTOPEXE): tools/tolink.cmx $(LINKCMX) $(LIBCOQRUN) $(TOPLOOPCMA:.cma=.cmxs) $(COQTOP_OPT_MLTOP)
 	$(SHOW)'COQMKTOP -o $@'
-	$(HIDE)$(COQMKTOP) -boot -opt $(OPTFLAGS) $(LINKMETADATA) -o $@
+	$(HIDE)$(OCAMLOPT) -linkall -linkpkg -I vernac -I toplevel -I tools \
+	                   -I kernel/byterun/ -cclib -lcoqrun \
+	                   -package num,str,unix,dynlink,threads,camlp5.gramlib \
```
Comment (gares) [inline Makefile.build]: This set of packages is also needed by coq_makefile:
   https://github.com/coq/coq/blob/master/tools/CoqMakefile.in#L101
Thanks to this PR we can now put this list of packages in config/Makefile and config/coq_config and be sure all users pick up the same value.

**Review by gares — DISMISSED** [review]
request: factor the list of packages linked by coq

**Comment by gares** [comment]
I like the patch.  The changes I require can also be seen as a future improvement.
I think this kind of PR should be merged early in the release cycle in order to be well tested (i.e. no human can proof read the makefile and be sure it is OK).

File: Makefile.build (hunk 1 above)
Comment (ejgallego) [reply Makefile.build]: I am not sure the set of packages is common, why should it be? I can imagine both packages having different linking sets.

**Comment by ejgallego** [comment]
I will do some refactoring thou; however my tendency is to specify toplevel link option by executable, and so they are easier to override.

**Comment by ejgallego** [comment]
I did a bit of refactoring on the packages linked, but however, I still think that the set of packages is a property of particular modules, in the form of dependencies. For example, in a `jbuilder`-based setup, you specify the dependencies for each subsystem, and a `META`, `merlin` etc... are generated automatically.

So I think this may be ready to be considered again.

File: configure.ml [hunk 2]
```diff
@@ -666,17 +666,15 @@ let natdynlinkflag =
...
 
-let osdeplibs = "-cclib -lunix"
-
-let operating_system, osdeplibs =
+let operating_system =
   if starts_with arch "sun4" then
     let os, _ = run "uname" ["-r"] in
     if starts_with os "5" then
-      "Sun Solaris "^os, osdeplibs^" -cclib -lnsl -cclib -lsocket"
+      "Sun Solaris "^os
```
Comment (psteckler) [inline configure.ml]: I was wondering if Solaris still existed. Yes, but: https://www.networkworld.com/article/3222707/data-center/the-sun-sets-on-solaris-and-sparc.html

File: Makefile.build [hunk 3]
```diff
@@ -379,32 +378,40 @@ coqbinaries: $(COQMKTOP) $(COQTOPEXE) $(CHICKEN) $(CSDPCERT) $(FAKEIDE)
```
Comment (vbgl) [inline Makefile.build]: It seems that a reference to the `COQMKTOP` variable has been forgotten, on line 377.

Comment (ejgallego) [reply Makefile.build]: Thanks @vbgl !

**Comment by herbelin** [comment]
There were users relying on `coqmktop` at some time. Are we sure there are none any more?

**Comment by ejgallego** [comment]
> There were users relying on coqmktop at some time. Are we sure there are none any more?

The only one I could think of is pidetop, but they'd be better served by using `ocamlfind` IMO.

**Comment by herbelin** [comment]
So, how to communicate to these users your opinion? Using a temporary deprecation with a little explanatory message?

**Comment by ejgallego** [comment]
So in fact CoqPIDEtop doesn't even use `coqmktop`. I am having trouble finding any user, and indeed I can't imagine why you would like to use that internal program as since ocamlfind become required, it is a much better alternative.

That being said, we could indeed deprecate it, but IMO not worth the effort.

**Comment by ejgallego** [comment]
Again I think the review is stale now @gares @maximedenes , will dismiss it soon.

I take way more seriously @herbelin concerns, but I think it is not a problem, but this kind of change to the build system is always delicate. With that in mind, I think we should go for it.

**Comment by Zimmi48** [comment]
If there were users at some point, there is still the risk that there were some users still, especially given that this was documented in the reference manual (for instance, I had read this section, even if I had never made use of this). So, if the replacement with `ocamlfind` is easy, maybe rather than just removing the section in the reference manual, you could replace it with a small explanation on how to achieve the same goals with `ocamlfind` and also add a mention in `CHANGES` about this removal.

**Comment by ejgallego** [comment]
I do agree, I will add a small example on how to use ocamlfind to build a coq toplevel.

**Comment by ejgallego** [comment]
Indeed, further improvements to this PR are at: https://github.com/ejgallego/coq/tree/mktop%2Bno_tolink

**Comment by psteckler** [comment]
I applied this to today's `master`, and got build errors.

With serial `make`:
```
OCAMLOPT  kernel/uGraph.ml
File "kernel/uGraph.ml", line 893, characters 14-38:
Error: Unbound module CProfile
```
With `make -j`:
```
OCAMLC    lib/coqProject_file.mli
make[1]: *** No rule to make target 'toplevel/coqtop_opt_bin.cmx', needed by 'bin/coqtop'.  Stop.
```

**Comment by ejgallego** [comment]
It looks like you didn't properly refresh your build env.

**Comment by psteckler** [comment]
> It looks like you didn't properly refresh your build env.

I did a `pull` and `clean` before building. What else?

**Comment by ejgallego** [comment]
What else could it be? Maybe the dependency info is not updated properly? Maybe you need a truly clean build env [=> `git clean -fxd` ?]

I dunno, it works well for me [just tested] .

**Comment by psteckler** [comment]
I tried again, seems to work now. Dunno what went wrong.

**Comment by SkySkimmer** [comment]
Did you reconfigure? Since this touches configure.ml

**Comment by psteckler** [comment]
> Did you reconfigure?

Yes. The problem might be that I ran `git clean` after patching and before local-committing, which would have removed new files.

**Pull request merged**